src/hotspot/share/opto/superword.hpp

```diff
@@ -46,331 +44,292 @@
+  bool transform_loop();
+
+  // VLoopAnalyzer
+  const VLoopAnalyzer& vla() const      { return _vla; }
+  IdealLoopTree* lpt() const            { return vla().lpt(); }
+  PhaseIdealLoop* phase() const         { return vla().phase(); }
+  PhaseIterGVN& igvn() const            { return vla().phase()->igvn(); }
+  CountedLoopNode* cl() const           { return vla().cl(); }
+  PhiNode* iv() const                   { return vla().iv(); }
+  int iv_stride() const                 { return vla().iv_stride(); }
+  bool in_body(const Node* n) const     { return vla().in_body(n); }
+
+  // VLoopAnalyzer reductions
+  bool is_marked_reduction(const Node* n) const {
+    return vla().reductions().is_marked_reduction(n);
+  }
+  bool reduction(Node* s1, Node* s2) const {
+    return vla().reductions().is_marked_reduction_pair(s1, s2);
+  }
 
-  bool transform_loop(IdealLoopTree* lpt, bool do_optimization);
+  // VLoopAnalyzer memory slices
+  bool same_memory_slice(MemNode* n1, MemNode* n2) const {
+    return vla().memory_slices().same_memory_slice(n1, n2);
+  }
 
-  void unrolling_analysis(int &local_loop_unroll_factor);
+  // VLoopAnalyzer body
+  const GrowableArray<Node*>& body() const {
+    return vla().body().body();
+  }
+  int body_idx(const Node* n) const     {
+    return vla().body().body_idx(n);
+  }
 
-  // Accessors for VPointer
-  PhaseIdealLoop* phase() const    { return _phase; }
-  IdealLoopTree* lpt() const       { return _lpt; }
-  PhiNode* iv() const              { return _iv; }
+  // VLoopAnalyzer dependence graph
+  bool independent(Node* s1, Node* s2) const {
+    return vla().dependence_graph().independent(s1, s2);
+  }
 
-  bool early_return() const        { return _early_return; }
+  // VLoopAnalyzer vector element type
+  const Type* velt_type(Node* n) const {
+    return vla().types().velt_type(n);
+  }
+  BasicType velt_basic_type(Node* n) const {
+    return vla().types().velt_basic_type(n);
+  }
+  bool same_velt_type(Node* n1, Node* n2) const {
+    return vla().types().same_velt_type(n1, n2);
+  }
+  int data_size(Node* n) const {
+    return vla().types().data_size(n);
+  }
+  int vector_width(Node* n) const {
+    return vla().types().vector_width(n);
+  }
+  int vector_width_in_bytes(Node* n) const {
+    return vla().types().vector_width_in_bytes(n);
+  }
 
 #ifndef PRODUCT
-  bool     is_debug()              { return _vector_loop_debug > 0; }
-  bool     is_trace_alignment()    { return (_vector_loop_debug & 2) > 0; }
-  bool     is_trace_mem_slice()    { return (_vector_loop_debug & 4) > 0; }
-  bool     is_trace_loop()         { return (_vector_loop_debug & 8) > 0; }
-  bool     is_trace_adjacent()     { return (_vector_loop_debug & 16) > 0; }
-  bool     is_trace_cmov()         { return (_vector_loop_debug & 32) > 0; }
+  // TraceAutoVectorization
+  bool is_trace_superword_adjacent_memops() const {
+    return vla().is_trace_superword_adjacent_memops();
+  }
```

Comment:
How about redefining it as:
```
  bool is_trace_superword_adjacent_memops() const {
    return TraceSuperWord || vla().is_trace_tag_active(TraceAutovectorizationTag::TAG_SW_ADJACENT_MEMOPS);
  }
```

And add a consulting interface in `class VLoop`:
```
  bool is_trace_tag_active(TraceAutovectorizationTag tag) const {
    return _trace_tags.at(tag);
  }
```

Thus, we don't have to involve any `SuperWord` specific words or options in shared facility.

Reply:
@fg1417 ok, I can do that :)